### PR TITLE
64 - ADR for single OAS document

### PR DESCRIPTION
**Pull request by ronytw**
Added the ADR for the single document.
Did not mention the Contract First mandate, as this can be in a different ADR entry.

### CODE REVIEW

File: docs/adr/0004-single-oas-document-for-all-the-vocabularies.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+## Decision
+
+After discussing with [Roberto](https://github.com/ioggstream), we have agreed on the following:
+
+* There will only be one OAS specification, which will apply for all of the different vocabularies. In other words, a
+  single OAS file that will look at lot
+  like [this one](https://editor.swagger.io/?url=https%3A%2F%2Fraw.githubusercontent.com%2Fioggstream%2Fjson-semantic-playground%2Fmaster%2Fopenapi%2Fvocabularies.yaml)
+  , with these differences:
+    * Instead of `vocabulary_id` we will have the pair `<agency_id, core_vocab_concept>`
```
Comment (ioggstream) [inline docs/adr/0004-single-oas-document-for-all-the-vocabularies.md:36]: @ronytw being an ADR, I'd generically  mention vocabulary and repository identifiers, and leave their name to the OAS specification file. This will allow to tweak them in case we need.

WRT the API, if we decide to have a single entrypoint (eg. so that agency_id or repository_id is a parameter) or multiple entrypoints (eg. one publishing container for every repo with agency_id/repository_id hardcoded in 

```
servers:
  url: https://repository_id.fqdn/vocabularies/v1/{...}
```
or
```
servers:
  - url: https://fqdn/vocabularies/v1/{repository_id}/{vocabulary_id}/ 
```

can be an implementation design choice.

Comment (ronytw) [reply docs/adr/0004-single-oas-document-for-all-the-vocabularies.md:36]: We're not planning to express any explicit mapping between containers and vocabularies. 
The implementation is going to be one for all the vocabularies and the identificators of agency / repo and vocabulary are of course parameters (whether they appear in the path or the query param has no implication on this).
How the load is distributed across instances of the backend implementation, is up to the container orchestrator. 
In other words, all the (code list service) backend instances can serve any request for any vocabularies, and they'll be replicated / scaled out as needed.

Comment (ioggstream) [reply docs/adr/0004-single-oas-document-for-all-the-vocabularies.md]: Ok. In general, sharding by domain or by path may impact on whether the scalability relies on the containre orchestrator only or on the architecture. Once we agree that this is out of scope for this ADR, it's ok.

Comment (ronytw) [reply docs/adr/0004-single-oas-document-for-all-the-vocabularies.md]: Yes, the scope of this ADR was only the OAS document. As we discussed yesterday, there will be a single OAS document that will describe the APIs for all of the vocabularies. Reasonably it will have two parametric `path` entries (get on the collection, get on the single item).

As we said if, for instance, there is a "education level" vocabulary from agency "agid", **there will not be an OAS document** which internally has the `/vocabularies/v1/agid/education-levels` path generated just for this API.
The single document will use [Path Templating](https://spec.openapis.org/oas/v3.1.0#path-templating) to refer to vocabulary identification tokens.

If you feel that we should make it explicit that we don't intend to implement sharding at domain name level for the MVP, we should maybe create a separate ADR entry.

Comment (ioggstream) [reply docs/adr/0004-single-oas-document-for-all-the-vocabularies.md]: > there will not be an OAS document which internally has the /vocabularies/v1/agid/education-levels path generated

That's correct.

> we should make it explicit that we don't intend to implement sharding at domain name level

No, since I generally like it :P I just mean that it's easier to postpone the decision for you :P

I think we can merge, then.
